lib/view_model/active_record/update_operation.rb

```diff
@@ -526,35 +540,121 @@ def build_updates_for_has_many_through(association_data, reference_strings, upda
       previous_through_viewmodels_by_indirect_ref =
         previous_through_viewmodels.group_by(&viewmodel_reference_for_indirect_model)
 
-      new_through_viewmodels = reference_strings.map do |ref|
-        target_reference = update_context.resolve_reference(ref).viewmodel_reference
+      case association_update
+      when ReferencedCollectionUpdate::Replace
+        target_indirect_viewmodels =
+          association_update.values.map do |ref|
+            {
+              ref:           ref,
+              viewmodel_ref: update_context.resolve_reference(ref).viewmodel_reference,
+            }
+          end
+
+      when ReferencedCollectionUpdate::Functional
+        target_indirect_viewmodels =
+          previous_through_viewmodels.map { |vm| { viewmodel_ref: viewmodel_reference_for_indirect_model.(vm) } }
+
+        functional_update_check_duplicates!(association_update, update_context)
+
+        association_update.values.each do |fupdate|
+          case fupdate
+          when FunctionalUpdate::Append
+            appended_targets = fupdate.values.map do |ref|
+              {
+                ref:           ref,
+                viewmodel_ref: update_context.resolve_reference(ref).viewmodel_reference
+              }
+            end
+
+            if fupdate.before || fupdate.after
+              moved_refs =
+                fupdate.values.map { |ref| update_context.resolve_reference(ref).viewmodel_reference }.to_set
+
+              target_indirect_viewmodels =
+                target_indirect_viewmodels.reject { |target| moved_refs.include?(target[:viewmodel_ref]) }
+
+              ref   = fupdate.before || fupdate.after
+              index = target_indirect_viewmodels.find_index { |target| target[:viewmodel_ref] == ref }
+              unless index
+                raise ViewModel::DeserializationError::NotFound.new(
+                  "Attempted to insert relative to reference that does not exist #{ref}",
+                  [ref])
+              end
+
+              index += 1 if fupdate.after
+              target_indirect_viewmodels.insert(index, *appended_targets)
+            else
+              target_indirect_viewmodels.concat(appended_targets)
+            end
+
+          when FunctionalUpdate::Remove
+            removed_refs = fupdate.values.to_set
+            target_indirect_viewmodels.reject! { |target| removed_refs.include? target[:viewmodel_ref] }
+
+          when FunctionalUpdate::Update
+            # Just need to connect these through, so the ref is anchored somewhere in the update tree.
+            # Uniquess of vm_refs has already been checked.=
+            ref_by_indirect_vm_ref =
+              fupdate.values.index_by { |ref| update_context.resolve_reference(ref).viewmodel_reference }
+
+            target_indirect_viewmodels.each do |target|
```

Comment:
So a `FU::Update` could also update something that was put into the `target_indirect_viewmodels` by a `FU::Append`? Or is this prevented by the reference uniqueness check at parse time?